README.md
# ember-getowner-polyfill

This provides a polyfill for the `Ember.getOwner` and `Ember.setOwner` features
added in Ember 2.3.

## Installation

```sh
ember install ember-getowner-polyfill
```

## Usage

```javascript
import Ember from 'ember';

export default Ember.Service.extend({
  someMethod() {
    let owner = Ember.getOwner(this);
    // do stuff with owner
  }
});
```

## Migration

### Applications

After you upgrade your application to Ember 2.3, you should remove `ember-getowner-polyfill` from
your `package.json`.

### Addons

Addons generally support many different Ember versions, so leaving `ember-getowner-polyfill` in
place for consumers of your addon is perfectly normal.  When the addon no longer supports Ember
versions older than 2.4, we recommend removing `ember-getowner-polyfill` from your `package.json`
and doing a major version bump.

## Documentation

The returned owner object should support all of the methods of the `ContainerProxyMixin` and `RegistryProxyMixin`.

* http://emberjs.com/api/classes/RegistryProxyMixin.html
* http://emberjs.com/api/classes/ContainerProxyMixin.html

## Compatibility

This addon is tested against quite a few past Ember versions. Check `config/ember-try.js` for the current list, but
the list of supported Ember versions at the time of authoring was:

* 1.10
* 1.11
* 1.12
* 1.13
* 2.0
* 2.1
* 2.2
* 2.3
* 2.4
* 2.8
* 2.12
* 2.14
* 2.15
* 2.16 (canary at the time)

## Addon Maintenance

### Installation

* `git clone` this repository
* `npm install`

### Running

* `ember serve`
* Visit your app at [http://localhost:4200](http://localhost:4200).

### Running Tests

* `npm test` (Runs `ember try:each` to test your addon against multiple Ember versions)
* `ember test`
* `ember test --server`

### Building

* `ember build`

For more information on using ember-cli, visit [https://ember-cli.com/](https://ember-cli.com/).


CHANGELOG.md
## v3.0.2 (2020-05-18)

#### :bug: Bug Fix
* [#35](https://github.com/ember-polyfills/ember-getowner-polyfill/pull/35) setOwner sets the container in Ember 1 ([@mmun](https://github.com/mmun))

#### Committers: 1
- Martin Muñoz ([@mmun](https://github.com/mmun))


## v3.0.1 (2020-05-14)

#### :bug: Bug Fix
* [#33](https://github.com/ember-polyfills/ember-getowner-polyfill/pull/33) Move ember-cli-htmlbars to devDependencies. ([@rwjblue](https://github.com/rwjblue))

#### Committers: 1
- Robert Jackson ([@rwjblue](https://github.com/rwjblue))


## v3.0.0 (2020-05-13)

#### :boom: Breaking Change
* [#32](https://github.com/ember-polyfills/ember-getowner-polyfill/pull/32) Update ember-cli blueprint from v2.16.2 to v3.18.0. ([@rwjblue](https://github.com/rwjblue))
* [#32](https://github.com/ember-polyfills/ember-getowner-polyfill/pull/32) Drop support for Node < 10. ([@rwjblue](https://github.com/rwjblue))

#### :rocket: Enhancement
* [#14](https://github.com/ember-polyfills/ember-getowner-polyfill/pull/14) Add `setOwner` support ([@jamesarosen](https://github.com/jamesarosen))

#### Committers: 2
- James A Rosen ([@jamesarosen](https://github.com/jamesarosen))
- Robert Jackson ([@rwjblue](https://github.com/rwjblue))




tests/test-helper.js
import QUnit from 'qunit';
import Ember from 'ember';
import Application from '../app';
import config from '../config/environment';
import { setApplication } from '@ember/test-helpers';
import { start } from 'ember-qunit';

setApplication(Application.create(config.APP));

let deprecations = [];

Ember.Debug.registerDeprecationHandler((message, options, next) => {
  deprecations.push(message);
  next(message, options);
});

QUnit.testStart(function() {
  deprecations = [];
});

QUnit.assert.noDeprecationsOccurred = function() {
  this.deepEqual(deprecations, [], 'Expected no deprecations during test.');
};

QUnit.assert.deprecations = function(callback, expectedDeprecations) {
  let originalDeprecations = deprecations;
  deprecations = [];

  callback();
  this.deepEqual(deprecations, expectedDeprecations, 'Expected deprecations during test.');

  deprecations = originalDeprecations;
};

start();


tests/unit/getowner-test.js
import Ember from 'ember';
import { moduleFor, test } from 'ember-qunit';
import hasEmberVersion from 'ember-test-helpers/has-ember-version';
const { getOwner } = Ember;

moduleFor('foo:bar', {
  beforeEach() {
    this.register('whatever:lol', Ember.Object.extend({
      isLOL: true
    }));
  },

  afterEach(assert) {
    assert.noDeprecationsOccurred();
  }
});

test('calling getOwner multiple times returns the same object', function(assert) {
  let subject = this.subject();
  assert.equal(getOwner(subject), getOwner(subject));
});

test('calling getOwner multiple times with different objects that share a container returns the same object', function(assert) {
  let subject = this.subject();

  let owner = getOwner(subject);
  owner.register('foo:baz', Ember.Object.extend({
    isBaz: true
  }));

  let result = owner.lookup('foo:baz');

  assert.strictEqual(getOwner(subject), getOwner(result));
});

test('it can use getOwner to look things up', function(assert) {
  let subject = this.subject();
  let owner = getOwner(subject);

  let result = owner.lookup('whatever:lol');

  assert.ok(result.isLOL, 'was able to lookup');
});

test('it can use getOwner to register things', function(assert) {
  let subject = this.subject();
  let owner = getOwner(subject);

  owner.register('foo:baz', Ember.Object.extend({
    isBaz: true
  }));

  let result = owner.lookup('foo:baz');

  assert.ok(result.isBaz, 'was able to register and lookup');
});

// In Ember 2.15 `_lookupFactory` was removed (in favor of `factoryFor`
if (!hasEmberVersion(2,15)) {
  test('it can use getOwner for the private _lookupFactory', function(assert) {
    let testFn = () => {
      let subject = this.subject();
      let owner = getOwner(subject);

      let Baz = Ember.Object.extend();
      Baz.reopenClass({ isBazFactory: true });

      owner.register('foo:baz', Baz);

      let result = owner._lookupFactory('foo:baz');

      assert.ok(result.isBazFactory, 'was able to register and _lookupFactory');
    };

    // we only can assert that a deprecation occurs when we are using
    // our custom `getOwner` polyfill. For Ember 2.3 - 2.11 we simply confirm
    // the functionality (not the deprecation).
    if (!hasEmberVersion(2,3) || hasEmberVersion(2,12)) {
      assert.deprecations(testFn, [
        'Using "_lookupFactory" is deprecated. Please use container.factoryFor instead.'
      ]);
    } else {
      testFn();
    }
  });
}

test('it can use getOwner for factoryFor', function(assert) {
  let subject = this.subject();
  let owner = getOwner(subject);

  let Baz = Ember.Object.extend();
  Baz.reopenClass({ isBazFactory: true });

  owner.register('foo:baz', Baz);

  let result = owner.factoryFor('foo:baz');

  assert.ok(result.class.isBazFactory, 'was able to register and factoryFor');
  let instance = result.create();
  assert.ok(instance instanceof Baz, 'factoryFor().create() results in a valid instance');
});

test('it can use hasRegistration', function(assert) {
  let subject = this.subject();
  let owner = getOwner(subject);

  let result = owner.hasRegistration('whatever:lol');

  assert.ok(result, 'hasRegistration works');
});

test('it can use inject', function(assert) {
  let subject = this.subject();
  let owner = getOwner(subject);

  owner.inject('whatever:lol', 'baz', 'foo:baz');
  owner.register('foo:baz', Ember.Object.extend({
    isBaz: true
  }));

  let result = owner.lookup('whatever:lol');

  assert.ok(result.get('baz.isBaz'), 'inject worked properly');
});

test('it can use registerOptionsForType', function(assert) {
  let subject = this.subject();
  let owner = getOwner(subject);

  owner.registerOptionsForType('lol', { instantiate: false });

  let result = owner.registeredOptionsForType('lol');

  assert.deepEqual(result, { instantiate: false }, 'registerdOptionsForType worked properly');
});

test('it can use resolveRegistration', function(assert) {
  let subject = this.subject();
  let owner = getOwner(subject);
  let Baz = Ember.Object.extend();

  owner.register('foo:baz', Baz);

  let result = owner.resolveRegistration('foo:baz');

  assert.equal(result, Baz, 'was able to resolveRegistration');
});

test('it can use unregister', function(assert) {
  let subject = this.subject();
  let owner = getOwner(subject);
  let Baz = Ember.Object.extend();
  let result;

  owner.register('foo:baz', Baz);

  result = owner.resolveRegistration('foo:baz');

  assert.equal(result, Baz, 'was able to resolveRegistration');

  owner.unregister('foo:baz');

  result = owner.resolveRegistration('foo:baz');

  assert.equal(result, undefined, 'was able to resolveRegistration');
});

test('it can use ownerInjection', function(assert) {
  let subject = this.subject();
  let owner = getOwner(subject);
  let baz = Ember.Object.create(owner.ownerInjection());
  let bazOwner = getOwner(baz);

  let result = bazOwner.lookup('whatever:lol');

  assert.ok(result.isLOL, 'was able to lookup');
});


tests/unit/setowner-test.js
import Ember from 'ember';
import EmberObject from '@ember/object';
import { getOwner, setOwner } from '@ember/application';
import { inject as service } from '@ember/service';
import { moduleFor, test } from 'ember-qunit';
import { lte } from 'ember-compatibility-helpers'

moduleFor('foo:bar', {
  afterEach(assert) {
    assert.noDeprecationsOccurred();
  }
});

test('setOwner works with getOwner', function(assert) {
  let object = {};
  let owner = {};

  assert.strictEqual(getOwner(object), undefined, 'pre-condition: object has no owner');

  setOwner(object, owner);

  assert.strictEqual(getOwner(object), owner, 'the owner is set correctly');
});

if (lte('1.13.13')) {
  test('[Ember 1.x]: setOwner sets the container if the owner has a container', function(assert) {
    this.register('service:session', Ember.Object);

    let Foo = EmberObject.extend({
      session: service()
    });

    let foo = Foo.create();

    setOwner(foo, getOwner(this));

    assert.ok(!!foo.get('session'), 'session injection exists');
  });

  test('[Ember 1.x]: setOwner does not crash if the owner does not have a container', function(assert) {
    this.register('service:session', Ember.Object);

    let Foo = EmberObject.extend({
      session: service()
    });

    let owner = {};
    let foo = Foo.create();

    setOwner(foo, owner);

    assert.throws(() => {
      foo.get('session');
    }, /Attempting to lookup an injected property on an object without a container/)
  });

  test('[Ember 1.x]: setOwner does not crash if the owner is undefined', function() {
    setOwner({}, undefined);
  });
}


index.js
'use strict';

var VersionChecker = require('ember-cli-version-checker');
var hasBeenWarned = false;

module.exports = {
  name: require('./package').name,
  included() {
    this._super.included.apply(this, arguments);

    this._ensureThisImport();

    var checker = new VersionChecker(this);
    var emberVersion = checker.forEmber();

    if (emberVersion.lt('2.3.0')) {
      this.import('vendor/install-getowner-polyfill.js');
    } else if (this.parent === this.project && !hasBeenWarned){
      console.warn('ember-getowner-polyfill is not required for Ember 2.3.0 and later, please remove from your `package.json`.');
      hasBeenWarned = true;
    }
  },

  _ensureThisImport() {
    if (!this.import) {
      this._findHost = function findHostShim() {
        var current = this;
        var app;
        do {
          app = current.app || app;
        } while (current.parent.parent && (current = current.parent));
        return app;
      };
      this.import = function importShim(asset, options) {
        var app = this._findHost();
        app.import(asset, options);
      };
    }
  },

  treeForVendor(rawVendorTree) {
    let babelAddon = this.addons.find(
      addon => addon.name === 'ember-cli-babel'
    );

    let transpiledVendorTree = babelAddon.transpileTree(rawVendorTree, {
      babel: this.options.babel,

      'ember-cli-babel': {
        compileModules: false,
      },
    });

    return transpiledVendorTree;
  },
};


.eslintrc.js
'use strict';

module.exports = {
  root: true,
  parser: 'babel-eslint',
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: 'module',
    ecmaFeatures: {
      legacyDecorators: true
    }
  },
  plugins: [ ],
  extends: [
    'eslint:recommended',
  ],
  env: {
    browser: true
  },
  rules: {},
  overrides: [
    // node files
    {
      files: [
        '.eslintrc.js',
        '.template-lintrc.js',
        'ember-cli-build.js',
        'index.js',
        'testem.js',
        'blueprints/*/index.js',
        'config/**/*.js',
        'tests/dummy/config/**/*.js'
      ],
      excludedFiles: [
        'addon/**',
        'addon-test-support/**',
        'app/**',
        'tests/dummy/app/**'
      ],
      parserOptions: {
        sourceType: 'script'
      },
      env: {
        browser: false,
        node: true
      },
      plugins: ['node'],
      rules: Object.assign({}, require('eslint-plugin-node').configs.recommended.rules, {
        // add your custom rules and overrides for node files here
      })
    }
  ]
};


addon/fake-owner.js
import Ember from 'ember';
let CONTAINER = `__${new Date()}_container`;
let REGISTRY = `__${new Date()}_registry`;

export default class FakeOwner {
  constructor(object) {
    this[CONTAINER] = object.container;

    if (Ember.Registry) {
      // object.container._registry is used by 1.11
      this[REGISTRY] = object.container.registry || object.container._registry;
    } else {
      // Ember < 1.12
      this[REGISTRY] = object.container;
    }
  }

  // ContainerProxyMixin methods
  //
  // => http://emberjs.com/api/classes/ContainerProxyMixin.html
  //
  lookup() {
    return this[CONTAINER].lookup(...arguments);
  }

  _lookupFactory() {
    return this[CONTAINER].lookupFactory(...arguments);
  }

  ownerInjection() {
    return {
      container: this[CONTAINER]
    };
  }

  // RegistryProxyMixin methods
  //
  // => http://emberjs.com/api/classes/RegistryProxyMixin.html
  //
  hasRegistration() {
    return this[REGISTRY].has(...arguments);
  }

  inject() {
    return this[REGISTRY].injection(...arguments);
  }

  register() {
    return this[REGISTRY].register(...arguments);
  }

  registerOption() {
    return this[REGISTRY].option(...arguments);
  }

  registerOptions() {
    return this[REGISTRY].options(...arguments);
  }

  registerOptionsForType() {
    return this[REGISTRY].optionsForType(...arguments);
  }

  registeredOption() {
    return this[REGISTRY].getOption(...arguments);
  }

  registeredOptions() {
    return this[REGISTRY].getOptions(...arguments);
  }

  registeredOptionsForType(type) {
    if (this[REGISTRY].getOptionsForType) {
      return this[REGISTRY].getOptionsForType(...arguments);
    } else {
      // used for Ember 1.10
      return this[REGISTRY]._typeOptions[type];
    }
  }

  resolveRegistration() {
    return this[REGISTRY].resolve(...arguments);
  }

  unregister() {
    return this[REGISTRY].unregister(...arguments);
  }
}


config/ember-try.js
'use strict';

module.exports = async function() {
  return {
    useYarn: true,
    scenarios: [
    {
      name: 'ember-1.10',
      bower: {
        dependencies: {
          "ember": "~1.10.0"
        }
      },
      npm: {
        devDependencies: {
          "ember-source": null
        }
      }
    },
    {
      name: 'ember-1.11',
      bower: {
        dependencies: {
          "ember": "~1.11.3"
        }
      },
      npm: {
        devDependencies: {
          "ember-source": null
        }
      }
    },
    {
      name: 'ember-1.12',
      bower: {
        dependencies: {
          "ember": "~1.12.1"
        }
      },
      npm: {
        devDependencies: {
          "ember-source": null
        }
      }
    },
    {
      name: 'ember-1.13',
      bower: {
        dependencies: {
          "ember": "~1.13.8"
        }
      },
      npm: {
        devDependencies: {
          "ember-source": null
        }
      }
    },
    {
      name: 'ember-2.0',
      bower: {
        dependencies: {
          "ember": "~2.0.0"
        }
      },
      npm: {
        devDependencies: {
          "ember-source": null
        }
      }
    },
    {
      name: 'ember-2.1',
      bower: {
        dependencies: {
          "ember": "~2.1.0"
        }
      },
      npm: {
        devDependencies: {
          "ember-source": null
        }
      }
    },
    {
      name: 'ember-2.2',
      bower: {
        dependencies: {
          "ember": "~2.2.0"
        }
      },
      npm: {
        devDependencies: {
          "ember-source": null
        }
      }
    },
    {
      name: 'ember-2.3',
      bower: {
        dependencies: {
          "ember": "~2.3.0"
        }
      },
      npm: {
        devDependencies: {
          "ember-source": null
        }
      }
    },
    {
      name: 'ember-2.4',
      bower: {
        dependencies: {
          "ember": "~2.4.0"
        }
      },
      npm: {
        devDependencies: {
          "ember-source": null
        }
      }
    },
    {
      name: 'ember-2.8',
      bower: {
        dependencies: {
          "ember": "~2.8.0"
        }
      },
      npm: {
        devDependencies: {
          "ember-source": null
        }
      }
    },
    {
      name: 'ember-2.12',
      npm: {
        devDependencies: {
          "ember-source": "~2.12.0"
        }
      }
    },
    ]
  };
};
